classes/module-signing.bbclass

```diff
@@ -17,6 +17,10 @@ export KERNEL_MODULE_SIG_KEY
 # build-time keypair will be generated and used for signing and embedding.
 export KERNEL_MODULE_SIG_CERT
 
+def get_signing_key(d):
+    path = d.getVar("KERNEL_MODULE_SIG_CERT") or os.path.join(d.getVar("STAGING_KERNEL_BUILDDIR"),"certs","signing_key.x509")
```

Comment:
I think `STAGING_KERNEL_BUILDDIR` should be `B` here.  Looking at oe-core:
`meta/classes/base.bbclass:do_fetch[file-checksums] = "${@bb.fetch.get_checksum_file_list(d)}"`
The file-checksums are the inputs to the task.  

do_compile will generate the key, so do_configure is too early (today it might inadvertantly be depending on an old key still in STAGING_KERNEL_BUILDDIR?)
do_shared_workdir can depend on B since it is ordered after do_compile, and it make sense since it is copying from B to STAGING_KERNEL_BUILDDIR.

All out of tree modules depend on do_shared_workdir, so they should be covered.

For in tree, the existing task deps are do_install -> do_sizecheck -> do_strip -> do_compile_kernelmodules -> do_shared_workdir, so that should already be covered by do_shared_workdir, I think.

Reply:
`do_configure` has to at least depend on `KERNEL_MODULE_SIG_CERT` and account if it is set or unset to update the defconfig.
Maybe then vardeps is more appropriate  for the defconfig and we can set `do_compile` to check `file-checksums` of either `KERNEL_MODULE_SIG_CERT` or `d.getVar("B"),"certs","signing_key.x509"`?

Reply:
Ah, good point about `do_configure` depending on KERNEL_MODULE_SIG_CERT being set or unset.  I just used bitbake-dumpsigs and verified OE already creates a dependency on the variable's value.

Here are the module signing relevant tasks:
```
do_configure:
  [ -n KERNEL_MODULE_SIG_CERT ] && update .config
do_compile
  [ -z KERNEL_MODULE_SIG_CERT ] && generate key/cert
  embed cert/public key
do_shared_workdir
  (implicitly [ -z KERNEL_MODULE_SIG_CERT ] ) cp ${B} cert to STAGING_KERNEL_BUILDDIR
do_install 
  [ -z KERNEL_MODULE_SIG_CERT ] && sign modules ( B for in-tree & STAGING_KERNEL_BUILDDIR for out-of-tree )
do_sign_module
  [ -n KERNEL_MODULE_SIG_KEY ] && sign modules
```

do_configure depends on KERNEL_MODULE_SIG_CERT's setting
do_compile depends on the cert's contents.  This is straightforward for KERNEL_MODULE_SIG_CERT, but tricky for an auto-generated key since the task also creates it.

That implies something like:
```
if KERNEL_MODULE_SIG_CERT
    do_compile[file-checksums] += "${@get_signing_cert(d)}"
else
    do_shared_workdir[file-checksums] += "${@get_autogen_signing_cert(d)}"
```

`get_*signing_key` with an empty return value as appropriate.

We could have a new task between do_configure and do_compile to call `make certs/signing_key.pem signing_key.x509` which would unify on do_compile[file-checksums].

This addition:
`do_sign_modules[file-checksums] += "${@get_signing_key(d)}"`
Yes, we need a dependency there, but I think it should be on KERNEL_MODULE_SIG_KEY and not the cert?

Reply:
Right, bitbake is able to pickup the vardep since it is explicit in the step. I'll make the changes and try that.

I think at this point the main issue is the sstate being used when we don't want it to.

Reply:
> I think at this point the main issue is the sstate being used when we don't want it to.

From https://bugzilla.yoctoproject.org/show_bug.cgi?id=14344
"So we don't support changing task hashes part way through the build."

This attempt to hash the signing key/cert doesn't work like we want.

Reply:
Yes.
I do not see any good way to both leverage the sstate and the auto-generated certificate reliably.
I think our options are:
- pass another certificate that was generated out of the tree (https://github.com/apertussolutions/bordel/pull/51, although we could re-use `repo-{dev,prod}` used to sign the XC-REPOSITORY stuff);
- use a test-signing certificate for development/debug purposes if none is specified out of tree (meta-security does that https://git.yoctoproject.org/cgit/cgit.cgi/meta-security/tree/meta-integrity/data/debug-keys);